### PR TITLE
server: get tikv config by status api

### LINKED ISSUE #4459: sync tikv config
## Enhancement Task
Config supports `region-max-size` property. This value should only be set by file not by api.

### PR DESCRIPTION
**Pull request by bufferflies**
Signed-off-by: bufferflies <1045931706@qq.com>

### What problem does this PR solve?

Issue Number: Close #4459

### What is changed and how it works?

PD should know the tikv configuration to schedule like merge_check.
```commit-message
```

### Check List

Tests

- Unit test

Code changes

Side effects

Related changes

### Release note

```release-note
None
```

### CODE REVIEW

File: server/config/store_config.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	if c == nil {
+		return value
+	}
+	body := fn()
+	var bs typeutil.ByteSize
+	if err := json.Unmarshal(body, &bs); err != nil {
+		return value
+	}
+	return bs.ToMiB()
+}
```
Comment (AndreMouche) [inline server/config/store_config.go:79]: It is too complex, it seems you just need a function to parse MBsize from a string ?

File: server/config/store_config.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	return c.getMbSize(fn, defaultRegionMaxSize)
+}
+
+// GetRegionSplitSize returns the region split size in MB
+func (c *StoreConfig) GetRegionSplitSize() uint64 {
+	fn := func() []byte {
+		return []byte(`"` + c.Coprocessor.RegionSplitSize + `"`)
+	}
+	return c.getMbSize(fn, defaultRegionSplitSize)
+}
```
Comment (AndreMouche) [inline server/config/store_config.go:95]: How about 
```
// GetRegionMaxSize returns the max region size in MB
func (c *StoreConfig) GetRegionMaxSize() uint64 {
	return parseMBFromTxt(c.Coprocessor.RegionMaxSize, defaultRegionMaxSize)

}

// GetRegionSplitSize returns the region split size in MB
func (c *StoreConfig) GetRegionSplitSize() uint64 {
	return parseMBFromTxt(c.Coprocessor.RegionSplitSize, defaultRegionSplitSize)
}

func parseMBFromTxt(text string, defaultSizeInMB uint64) uint64 {
	b := typeutil.ByteSize(0)
	err := b.UnmarshalText([]byte(text))
	if err != nil {
		return defaultSizeInMB
	} else {
		return b.ToMiB()
	}
}


```

Comment (bufferflies) [reply server/config/store_config.go:95]: yes, I will fix it.

File: server/grpc_service.go [hunk 3]
```diff
@@ -522,6 +522,14 @@ func (s *GrpcServer) PutStore(ctx context.Context, request *pdpb.PutStoreRequest
 	log.Info("put store ok", zap.Stringer("store", store))
 	CheckPDVersion(s.persistOptions)
 
+	go func(url string) {
+		// tikv may not ready to server.
```
Comment (lhy1024) [inline server/grpc_service.go:526]: What happens to down tikv?

Comment (bufferflies) [reply server/grpc_service.go:526]: Keeping  the config of all the tikv  to be same in the cluster is the precondition，so the pd should be sync the tikv config even through some tikv is down.  if the precondition is not exist, the cluster will work well in past.
 In upgrade, the configuration of the TiKV may be different,syncing with tikv is ok in the end.

**Comment by bufferflies** [comment]
PTAL @lhy1024 @AndreMouche again.

**Comment by bufferflies** [comment]
PTAL @AndreMouche @lhy1024 again.

**Comment by rleungx** [comment]
I'm worrying that we had lots of ways to support saving TiKV configs in PD before and all of them are deprecated. Those ways are more general than this PR. So I think we need a further discussion of it before moving forward.

**Review by AndreMouche — APPROVED** [review]
LGTM

**Comment by bufferflies** [comment]
> I'm worrying that we had lots of ways to support saving TiKV configs in PD before and all of them are deprecated. Those ways are more general than this PR. So I think we need a further discussion of it before moving forward.

I I very much agree with your opinion about the consistency with other component,  it does need a further discussion. But in past, this is the smallest cost to get the config of the tikv and tikv doesn't need to add other codes. it will need to rewrite the update logic.  I will propose one issue about the centralized configuration,.

File: server/grpc_service.go (hunk 3 above)
Comment (disksing) [reply server/grpc_service.go:527]: Does TiFlash have the status API too?

Comment (bufferflies) [reply server/grpc_service.go:527]: it is a good question, it should ignore tiflash.

**Review by disksing — APPROVED** [review]

File: pkg/typeutil/size_test.go [hunk 4]
```diff
@@ -16,7 +16,6 @@ package typeutil
 
 import (
 	"encoding/json"
-
```
Comment (lhy1024) [inline pkg/typeutil/size_test.go:19]: unnecessary modify

File: server/config/store_config.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+// NewStoreConfigManager creates a new StoreConfigManager.
+func NewStoreConfigManager(config *SecurityConfig) *StoreConfigManager {
+	manager := &StoreConfigManager{
+		schema: "http",
+	}
+	if config == nil {
+		return manager
+	}
+	if cfg, err := config.ToTLSConfig(); err == nil && cfg != nil {
```
Comment (lhy1024) [inline server/config/store_config.go:56]: Do we need to add a tls test?

File: server/config/store_config.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+// StoreConfig is the config of store like TiKV.
+// generated by https://mholt.github.io/json-to-go/.
+// nolint
+type StoreConfig struct {
+	Coprocessor `json:"coprocessor"`
+}
+
+// Coprocessor is the config of coprocessor.
+type Coprocessor struct {
+	RegionMaxSize   string `json:"region-max-size"`
```
Comment (lhy1024) [inline server/config/store_config.go:74]: how about using atomic value to avoid introducing lock?

File: server/schedule/checker/merge_checker.go [hunk 7]
```diff
@@ -32,7 +32,10 @@ import (
 	"github.com/tikv/pd/server/schedule/placement"
 )
 
-const maxTargetRegionSize = 500
+const (
+	maxTargetRegionSize   = 500
+	maxTargetRegionFactor = 4
```
Comment (lhy1024) [inline server/schedule/checker/merge_checker.go:37]: why use it?

File: tests/server/cluster/cluster_test.go [hunk 8]
```diff
@@ -660,7 +660,7 @@ func (s *clusterTestSuite) TestLoadClusterInfo(c *C) {
 	tc.WaitLeader()
 	leaderServer := tc.GetServer(tc.GetLeader())
 	svr := leaderServer.GetServer()
-	rc := cluster.NewRaftCluster(s.ctx, svr.ClusterID(), syncer.NewRegionSyncer(svr), svr.GetClient(), svr.GetHTTPClient())
+	rc := cluster.NewRaftCluster(s.ctx, svr.ClusterID(), syncer.NewRegionSyncer(svr), svr.GetClient(), svr.GetHTTPClient(), svr.GetStoreConfigManager())
```
Comment (lhy1024) [inline tests/server/cluster/cluster_test.go:663]: I think we need to add a test for modifying config

File: server/config/store_config.go (hunk 6 above)
Comment (bufferflies) [reply server/config/store_config.go:74]: good idea.

File: tests/server/cluster/cluster_test.go (hunk 8 above)
Comment (bufferflies) [reply tests/server/cluster/cluster_test.go:663]: store manage expose the  `UpdateConfig` api.

File: pkg/typeutil/size_test.go (hunk 4 above)
Comment (bufferflies) [reply pkg/typeutil/size_test.go]: fix

File: server/config/store_config.go (hunk 5 above)
Comment (bufferflies) [reply server/config/store_config.go:56]: yes , fixed

**Review by lhy1024 — APPROVED** [review]

**Pull request merged**